### PR TITLE
[HGCAL] Pattern Recognition as Plugins, with CLUE3D addition.

### PR DESCRIPTION
**Pull request by rovere**
#### PR description:

This PR will move the current Pattern Recognition algorithms in `HGCAL` to become plugins.
In addition to that, a new Pattern Recognition algorithm, `CLUE3D`, has been added, as a plugin. For the time being, it's not run in production, but it has been tested.
Pattern Recognition parameters have been moved directly into the corresponding plugins.

Expect no regression.

#### PR validation:

Typical single-particle workflows related to HGCAL. Tested enabling and disabling `CLUE3D`.

### CODE REVIEW

**Comment by rovere** [comment]
@cmsbuild please test

File: RecoHGCal/TICL/plugins/TrackstersProducer.cc [hunk 1]
```diff
@@ -92,9 +90,17 @@ TrackstersProducer::TrackstersProducer(const edm::ParameterSet& ps, const Tracks
           consumes<std::vector<TICLSeedingRegion>>(ps.getParameter<edm::InputTag>("seeding_regions"))),
       itername_(ps.getParameter<std::string>("itername")) {
   if (doNose_) {
+    auto plugin = ps.getParameter<std::string>("patternRecognitionBy");
```
Comment (lecriste) [inline RecoHGCal/TICL/plugins/TrackstersProducer.cc:93]: This and next line can be moved out of the `if`, removing their repetiion in the `else`.

File: RecoHGCal/TICL/python/CLUE3DHighStep_cff.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ticlTrackstersCLUE3DHigh = _trackstersProducer.clone(
+    filtered_mask = "filteredLayerClustersCLUE3DHigh:CLUE3DHigh",
+    seeding_regions = "ticlSeedingGlobal",
+    itername = "CLUE3DHigh",
+    patternRecognitionBy = "CLUE3D"
+)
+
+# MULTICLUSTERS
+
+ticlMultiClustersFromTrackstersCLUE3DHigh = _multiClustersFromTrackstersProducer.clone(
```
Comment (lecriste) [inline RecoHGCal/TICL/python/CLUE3DHighStep_cff.py:28]: Do we still want to produce these MultiClusters?

File: RecoHGCal/TICL/python/CLUE3DLowStep_cff.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    itername = "CLUE3DLow",
+    patternRecognitionBy = "CLUE3D",
+    pluginPatternRecognitionByCLUE3D = dict (
+        criticalDensity = 2.
+    )
+)
+
+# MULTICLUSTERS
+
+ticlMultiClustersFromTrackstersCLUE3DLow = _multiClustersFromTrackstersProducer.clone(
```
Comment (lecriste) [inline RecoHGCal/TICL/python/CLUE3DLowStep_cff.py:32]: As above, do we still want to produce these MultiClusters?

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> `CLUE3D`, has been added, as a plugin


please add a link to some slides with more detail on the plans for this algorithm.
In the PR there are two new iterations based on `CLUE3D` and they are no longer arranged by physics purpose of the iteration (like EMStep) and instead have `CLUE3DHigh` and `CLUE3DLow`.
Is this a temporary setup and the eventual goal is to replace the implementation of individual EM/Trk/Had/etc or will it really replace all other EM/Trk/Had/etc?

If the goal is to swap-out individual EM/Trk/Had iterations, it may be more practical to define modifiers.
If the goal is to use CLUE3D concurrently, is there a better physics-like naming compared to  `CLUE3DHigh` and `CLUE3DLow`

**Review by slava77 — COMMENTED** [review]
please provide a test configuration to check CLUE3D in runtime

**Comment by rovere** [comment]
> > `CLUE3D`, has been added, as a plugin
> 
> please add a link to some slides with more detail on the plans for this algorithm.
> In the PR there are two new iterations based on `CLUE3D` and they are no longer arranged by physics purpose of the iteration (like EMStep) and instead have `CLUE3DHigh` and `CLUE3DLow`.
> Is this a temporary setup and the eventual goal is to replace the implementation of individual EM/Trk/Had/etc or will it really replace all other EM/Trk/Had/etc?
> 
> If the goal is to swap-out individual EM/Trk/Had iterations, it may be more practical to define modifiers.
> If the goal is to use CLUE3D concurrently, is there a better physics-like naming compared to `CLUE3DHigh` and `CLUE3DLow`

Ciao Slava,
thanks for the questions. The use cases for `CLUE3D` (all to be studied, tuned and verified) could be many. The main idea, at present, is to use this first pass of Pattern Recognition to identify high-energy seeding regions (that could be themselves Tracksters) and use those regions to then trigger a second Pattern Recognition, CA-based, with much more insight on what to expect (and what to look for starting from where), unlike the present case for the unseeded iterations.
If `CLUE3D` proves to be "good enough", it could even be promoted to be an iteration (or many) by itself, producing not just seeds, but Tracksters.
The current naming, `High` and `Low`, refers to the density required to create a seed (in the algorithm, not in TICL terms): maybe adding `Density` explicitly, could help to understand that.
I hope this clarifies a bit more about what we are aiming at. We will be happy to extend, in case you have more specific questions.
I'll review the code-related comments inline, later on.

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by rovere** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> Pull request #33645 was updated.

I was just starting with the follow-up review.

@rovere 
please clarify if you are done with updates or if something is still pending.
Thank you.

**Comment by rovere** [comment]
@slava77 
Done, thanks.

**Comment by rovere** [comment]
Ciao,
is there any hope this PR would enter `pre2`?

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33645	d501c9e
- code changes are in line with the PR description and reasonably follow the code review
- jenkins tests pass and comparisons with the baseline show no (relevant) differences as expected for the part of the PR where the currently running code is refactored/reorganized in a mostly technical way
- a local test with wf 34834.999 (PU50 phase-2) and clue3D process modifier (also with relevant `keep` for the output tracksters) show that the High and Low variants of CLUE3D run to completion with somewhat reasonable but perhaps still to be tuned/reduced time of 220-230 ms/evt each (vs e.g. 107 ms/evt in ticlTrackstersEM)

**Comment by srimanob** [comment]
+Upgrade

This PR is to organize the PR algorithms to become plugin, and introduce new CLUE3D PR. No change is expected on wf as of now.

**Comment by silviodonato** [comment]
+1

**Pull request merged**